src/enzyme_ad/jax/Utils.cpp

```diff
@@ -566,6 +574,284 @@ bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy,
   return allowOnFloatingPointMath;
 }
 
+bool canApplyNoNanPattern(bool allowOnFloatingPointMath, Type outTy, Type inTy,
+                          mlir::Operation *op) {
+  outTy = getElementTypeOrSelf(outTy);
+  inTy = getElementTypeOrSelf(inTy);
+  if (outTy.isInteger() && inTy.isInteger())
+    return true;
+  return allowOnFloatingPointMath || guaranteedNoNanResult(op);
+}
+
+bool elementwiseConstantOpCheck(
+    stablehlo::ConstantOp constOp,
+    std::function<bool(DenseElementsAttr)> floatCheck,
+    std::function<bool(DenseElementsAttr)> intCheck) {
+  Attribute attr = constOp.getValue();
+
+  if (auto denseAttr = dyn_cast<DenseElementsAttr>(attr)) {
+    if (denseAttr.getType().getShape().size() && denseAttr.isSplat()) {
+      denseAttr = denseAttr.resizeSplat(
+          RankedTensorType::get({}, denseAttr.getType().getElementType()));
+    }
+
+    // For floating point values
+    if (isa<FloatType>(denseAttr.getElementType())) {
+      return floatCheck(denseAttr);
+    }
+
+    // For integer values
+    if (isa<IntegerType>(denseAttr.getElementType())) {
+      return intCheck(denseAttr);
+    }
+  }
+
+  return false;
+}
+
+bool guaranteedNoNanResult(stablehlo::ConstantOp constOp) {
+  return elementwiseConstantOpCheck(
+      constOp,
+      [](DenseElementsAttr attr) {
+        return !std::any_of(attr.getValues<APFloat>().begin(),
+                            attr.getValues<APFloat>().end(),
+                            [](APFloat elem) { return elem.isNaN(); });
+      },
+      [](DenseElementsAttr attr) { return true; });
+}
+
+bool guaranteedNoNanResult(mlir::Value value) {
+  return guaranteedNoNanResult(value.getDefiningOp());
+}
+
+bool guaranteedNoNanResult(mlir::Operation *op) {
+  if (!op)
+    return false;
+
+  if (auto constantOp = dyn_cast<mlir::stablehlo::ConstantOp>(op)) {
+    return guaranteedNoNanResult(constantOp);
+  }
+
+  // data movement ops
+  if (isa<stablehlo::SliceOp, stablehlo::ConcatenateOp,
+          stablehlo::BroadcastInDimOp, stablehlo::ReshapeOp,
+          stablehlo::TransposeOp>(op)) {
+    return std::all_of(op->getOperands().begin(), op->getOperands().end(),
+                       [](mlir::Value v) { return guaranteedNoNanResult(v); });
+  }
+
+  // integer ops
+  if (isa<stablehlo::AndOp, stablehlo::OrOp, stablehlo::XorOp, stablehlo::NotOp,
+          stablehlo::IotaOp>(op)) {
+    return true;
+  }
+
+  // elementwise ops that are no-nan if all operands are not nan
+  if (isa<stablehlo::AddOp, stablehlo::SubtractOp, stablehlo::AbsOp,
+          stablehlo::ExpOp, stablehlo::ConvertOp, stablehlo::CompareOp,
+          stablehlo::TanhOp, stablehlo::LogisticOp, stablehlo::FloorOp,
+          stablehlo::CeilOp>(op)) {
+    return std::all_of(op->getOperands().begin(), op->getOperands().end(),
+                       [](mlir::Value v) { return guaranteedNoNanResult(v); });
```

Comment:
add and sub are still problematic here, (inf + (-inf)), and inf - inf as examples.